[1단계 방탈출 결제 / 배포] 레오(김준영) 미션 제출합니다.

kjyyjk opened the pull request:
<!-- 

## 코드 리뷰 팁

- 코드와 관련된 질문이 있다면, PR 본문에 적기 보다는 해당 코드를 선택하고 코멘트를 남겨주세요.
  - [참고: Adding comments to a pull request](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/reviewing-changes-in-pull-requests/commenting-on-a-pull-request#adding-comments-to-a-pull-request)

-->
## 체크 리스트
- [x] 미션의 필수 요구사항을 모두 구현했나요?
- [x] Gradle `test`를 실행했을 때, 모든 테스트가 정상적으로 통과했나요?
- [x] 애플리케이션이 정상적으로 실행되나요?
- [x] [프롤로그](https://prolog.techcourse.co.kr)에 셀프 체크를 작성했나요?
https://prolog.techcourse.co.kr/studylogs/4439
  <!-- 작성한 셀프 체크의 링크를 남겨주세요. -->
## 객체지향 생활체조 요구사항을 얼마나 잘 충족했다고 생각하시나요?
### 1~5점 중에서 선택해주세요.

- [ ] 1 (전혀 충족하지 못함)
- [ ] 2
- [x] 3 (보통)
- [ ] 4
- [ ] 5 (완벽하게 충족)

### 선택한 점수의 이유를 적어주세요.
<!-- 이유 작성 -->
다른 부분들이 어떻게 구현되어있는지 아직 잘 파악하지 못하였기 때문에 보통을 선택했습니다.

## 베이스 코드 선택 체크
- [ ] 이전 미션의 내 코드에서 시작 
- [x] 이전 미션의 페어의 코드에서 시작

## 어떤 부분에 집중하여 리뷰해야 할까요?
<!-- 리뷰어가 효과적으로 피드백할 수 있도록 중점적으로 리뷰 받고 싶은 내용을 *요약* 형태로 작성해 주세요.
리뷰해야 할 포인트를 요약해 강조해 주시면, 리뷰어가 코드 전체를 이 부분에 집중해 리뷰할 수 있습니다.
반면 특정 코드부분에 대한 피드백이 필요하다면, 이 곳에 적기 보다 해당 코드를 선택하고 코멘트를 남기는 것을 권장합니다. -->

안녕하세요 파랑,
마지막 리뷰어로 만나 뵙게되어 반갑습니다~!👍

페어의 코드가 한 번도 경험해보지 못한 구조와 설계를 가지고 있어, 한 번 경험해보고 싶은 마음으로 페어 코드로 시작하기를 선택했습니다. 다만 이번 미션 요구사항에 집중하느라 그 외 기능들이 어떻게 구현되어있는지 아직 절반도 파악하지 못한 것 같아요.

그렇기 때문에 가능하다면 이번 리뷰에서는 결제 API 연동 부분과 해당 테스트를 위주로 봐주셨으면 좋겠습니다!

이번 미션 커밋입니다.
https://github.com/woowacourse/spring-roomescape-payment/pull/241/files/035661a73c86bb7cdfe01287b6d00517b2e68395..a4d1932cd8c0c90f4bcb2beabe55a001a175933f

이번 미션 리뷰 잘부탁드립니다.😊

## 애플리케이션 실행 관련
애플리케이션 실행 시 사용할 수 있는 admin과 user 계정을 등록해두었습니다. 로그인 정보는 다음과 같습니다.
```
admin
id : admin@email.com
password : qwe123

user1
id : user1@email.com
password : qwe123

user2
id : user2@email.com
password : qwe123
```

또한 민감 정보를 git에 올리지 않게 해두었습니다. 혹시 실행해보신다면 다음과 같이 yml을 추가하셔야합니다!

resources/application-auth.yml
```
jwt:
  key: /VCEkdrvbjkCLRp0G0LK12VsDYViO/kufd8UVphCq80MdiZzNlefqOcwqfpvW+jIRNBmhiNKbNUGXSwum0DSgUjkEl2To/350R/ffncBkVeO7tfNq/4rNXWGGHre9p8s14m0k2z/QS4=
  expiration: 360000

pay:
  toss:
    secret-key: test_gsk_docs_OaPz8L5KdmQXkzRz3y47BMw6
```

### 📌 코드리뷰 질문 그라운드룰 (지우지 마세요)
✅ 많은 사람과 공유하고 토론해보는 것이 더 좋은 질문에는, 리뷰어가 💬말풍선 이모지를 남겨 토론을 추천할 수 있어요.  
✅ 토론을 추천받은 질문은 `#be-7기-잡담-질답`슬랙채널이나 캠퍼스에서 자유롭게 토론을 이어가보세요!  
✅ 코드에 대한 맥락 없이 리뷰어의 주관적인 의견을 묻거나, 단순히 현업에서는 어떻게 하는지를 묻는 질문은 코드 리뷰 본문에는 어울리지 않아요.  
코드 리뷰는 변경된 코드의 목적, 의도, 품질을 중심으로 논의하는 자리예요.  
✅ 코드와 관련된 질문이 있다면 PR 본문에 적기보다는, 해당 코드를 선택해 코멘트를 남겨주세요.

unh6unh6 committed:
init: 이전 미션 코드 마이그레이션

unh6unh6 committed:
feat: TossPayments 결제 위젯 연동 및 결제 기반 예약 로직 구현

unh6unh6 committed:
feat: 결제 승인 API 호출 구현

unh6unh6 committed:
feat: 결제 승인 API 호출 구현

unh6unh6 committed:
feat: 결제 승인 API 호출 구현

# Conflicts:
#	src/main/java/roomescape/reservation/controller/ReservationController.java

kjyyjk committed:
Merge branch 'step1' of https://github.com/unh6unh6/spring-roomescape-payment into step1

unh6unh6 committed:
refactor: 결제 승인 연동 로직을 PaymentService에서 하도록 변경

unh6unh6 committed:
feat: 결제 예외 처리 로직 추가 및 관련 클래스 구현

kjyyjk committed:
Merge branch 'step1' of https://github.com/unh6unh6/spring-roomescape-payment into step1

kjyyjk committed:
feat: 결제 오류 처리 로직 및 예외 처리 개선

kjyyjk committed:
test: PaymentService 예외 처리 테스트

unh6unh6 committed:
feat: PaymentService에 요청 타임아웃 설정 추가 (connectTimeout, readTimeout)

unh6unh6 committed:
refactor: RestClient 설정 분리 및 RestClientConfig 추가

kjyyjk committed:
refactor: 미사용 import 제거

kjyyjk commented on src/test/java/roomescape/payment/PaymentServiceTest.java:
```diff
@@ -0,0 +1,69 @@
+    private PaymentService paymentService;
+
+    @Autowired
+    MockRestServiceServer mockServer;
```
본 테스트의 목적은 외부 서버로부터 예외 응답이 내려왔을 때, 우리의 애플리케이션이 잘 처리하는지 검증하는 것입니다.
그 과정에서 실제 외부 서버를 사용하는 것은 불필요하다고 판단하여
다음과 같이 `@RestClientTest`, `MockRestServiceServer`를 활용했습니다.

제가 직접 테스트를 작성하며 이해한 바는 다음과 같습니다.
`@RestClientTest`를 사용하면 RestClient.Builder와 이를 주입받는 컴포넌트들을 컨텍스트로 로드합니다.
또한 MockRestServiceServer를 자동 구성해 Builder로부터 생성된 RestClient에서 나가는 요청들을 가로채
mocking할 수 있게 합니다.

kjyyjk replied:
> 실제 외부 서버를 사용하는 것은 불필요하다고 판단하여

라고 말은 했지만 실제 서버를 찔러보는 테스트 또한 추가로 필요하다고 저는 생각합니다.

이에 대해 `외부 서버 상태에 따라 테스트가 실패할 수도 있기 때문에 굳이 싶다`라는 의견도 있었습니다.
해당 의견에도 일부 동의하지만 오히려 테스트 실패를 먼저 마주함으로써 더 큰 장애를 막을 수 있다고 생각합니다.
그것이 테스트의 역할이니까요!

만약 서비스의 모든 테스트를 MockServer만 사용해서 작성해놓는다면 외부 서버의 api 명세에 변경이 생겼다던가,
장애가 발생했을 때 안전하게 대응하지 못할 것 같다는 생각입니다.

파랑은 이에 대해 어떻게 생각하시는지 궁금합니다!

summerlunaa replied:
> 실제 서버를 찔러보는 테스트 또한 추가로 필요하다고 저는 생각합니다.

동의합니다. 처음 API 연동 기능을 개발한다면(혹은 수정이 생긴다면) 어떠한 방식으로든 실제로 호출을 해봐야 제대로 동작하는지 정확한 확인이 가능하겠죠. 하지만 테스트는 언제든 몇번이든 반복해서 실행될 수 있는데 외부 API 테스트도 동일하게 매번 실행되어야 하는가? 생각해보면 그건 또 아니거든요. 외부 서버의 상태 때문에 테스트가 깨질 수도 있고 호출 횟수에 제한이 있을 수도 있으니까요.
이 둘 모두를 만족하도록 테스트를 작성하는 방법을 고민해보셨을까요? 레오는 외부 API를 어떤 방식으로 테스트하는 게 좋다고 생각하는지 궁금합니다 :)

kjyyjk replied:
아뇨, 구체적인 방법은 고려해보지 않았었습니다.

다만 실제 결제 승인 API를 호출한다고 해도 정상 동작하는지 확인하려면 유효한 PaymentKey가 필요한데,
생각해보니 테스트 환경에서 유효한 PaymentKey를 얻기가 힘들겠네요..!

그렇다면 유효하지 않은 PaymentKey로 요청했을 때 403 Forbidden과 그에 맞는 예외 코드/메세지가
응답 오는지를 검증할 수 있을 것 같아요.
해당 테스트는 외부 API 서버가 살아있으며 응답까지 잘 내려준다는 것을 검증한다가 목적일 것 같습니다.

그리고 이처럼 실제 서버에 요청하는 테스트가 다른 테스트와 동일하게 매번 실행되는 것은 적절하지 않다는 의견에 동의합니다!

이에 대해서는 해당 테스트만 `@Disabled`처리 해두면 매번 같이 실행되지 않고 필요한 시점에서만 별개로 실행하는 방법도 
생각해볼 수 있을 것 같아요.

summerlunaa replied:
넵넵ㅎㅎㅎ 테스트 환경에서 실제 운영되는 서버를 직접 호출해서 테스트하는 것은 굉장히 제약이 많습니다. 만약 외부 API에서 테스트용 서버와 key를 제공한다면 테스트를 작성해볼수는 있겠죠. 그렇게 외부 API를 호출하는 테스트를 작성한다고 하더라도 매번 실행되는 것은 적절하지 않기 때문에 말씀하신 것처럼 `@Disabled` 처리해두고 필요 시에만 직접 테스트를 실행해보면 됩니다! 잘 고민해보셨네요 👍

kjyyjk commented on src/main/java/roomescape/common/config/RestClientConfig.java:
```diff
@@ -0,0 +1,19 @@
+@Configuration
+public class RestClientConfig {
+
+    @Bean
+    public RestClient.Builder restClientBuilder() {
+        final SimpleClientHttpRequestFactory factory = new SimpleClientHttpRequestFactory();
+        factory.setConnectTimeout(1500);
+        factory.setReadTimeout(1500);
```
외부 서버와의 연결 혹은 응답이 무한정 늦어질 경우를 대비해 TimeOut을 지정해주었습니다.

아직 api 호출 소요 시간에 대한 저의 기준이 존재하지 않아 우선 1.5초로 지정해두었습니다.

하지만 너무 짧은 거 아닌가하는 생각이 드네요.. 흠
단순히 토스 서버의 응답 속도 외에도 네트워크에 따른 속도 지연을 고려한다면 더 넉넉히 잡아야하지 않을까 싶습니다.

summerlunaa replied:
보통 readTimeout 보다는 connectionTimeout 시간을 짧게 잡습니다. 어느 정도가 적당할지 한번 고민해보세요~ 너무 짧으면 실패가 많아질 거고 너무 길면 응답이 지연되어 사용자가 불편함을 겪을 거예요.

kjyyjk replied:
토스 페이먼츠의 공식 문서에서는 ConnectionTimeOut 시간으로 30~60초가 적절하다고 안내하고 있습니다.
하지만 단순히 공식 문서를 따르기 보다는 사용자의 경험이 우선되어야할 것 같다고 판단했습니다.
제가 방탈출 예약 시스템의 사용자라고 상상해봤을 때, 결제 요청이 10초 이상 지연된다면 요청을 취소하거나 브라우저(혹은 앱)을 껐다 킬 것 같아요.
(인내심이 부족하기는 합니다..ㅎ)

그래서 ReadTimeOut = 8초, ConnectionTimeOut = 2초로 수정했습니다.
cff09631b1bfdf92264bb56d95ccaa1b25d02c79

추가로 ConnectionTimeOut은 보통 방화벽 설정이나 외부 서버 이상으로 발생하는 문제이기 때문에 짧게 여러 번 연결을 시도하는 것 또한 좋은 방식이라고 들었습니다.
이 경우에는 `@Retryable`을 활용해 n번 재시도하도록 설정해 볼 수 있을 것 같습니다.

summerlunaa replied:
> 하지만 단순히 공식 문서를 따르기 보다는 사용자의 경험이 우선되어야할 것 같다고 판단했습니다.

너무 좋습니다 👍 저도 30-60초는 너무 길다고 생각했습니다. Retryable까지 알아보셨네요! 모든 API가 재시도에 적합한 것은 아닙니다. API가 멱등성을 보장하는지(여러번 호출해서 결과가 달라진다면 중복 처리의 위험이 있겠죠), 어떤 케이스에만 재시도가 필요할지, 얼만큼의 간격으로 몇 번이나 재시도를 할지 등을 고민해볼 수 있을 것 같아요. 이번 미션의 범위는 아니니 이런 부분도 고려해야하는 구나~ 정도로 이해하고 넘어가시면 될 것 같습니다 :)

kjyyjk commented on src/main/java/roomescape/payment/PaymentErrorHandler.java:
```diff
@@ -0,0 +1,45 @@
+                    .anyMatch(tossErrorCode -> tossErrorCode.name().equals(code));
+        }
+    }
+}
```
https://docs.tosspayments.com/reference/error-codes
토스 결제 승인 API의 에러 코드가 정말 많더라고요..!
```
INVALID_API_KEY(잘못된 시크릿키 연동 정보 입니다.)
UNAUTHORIZED_KEY(인증되지 않은 시크릿 키 혹은 클라이언트 키 입니다.)
INCORRECT_BASIC_AUTH_FORMAT(잘못된 요청입니다. ':' 를 포함해 인코딩해주세요.)
```
이 3개의 에러 코드는 클라이언트 및 사용자가 몰라도 되는 정보라고 판단해 500 서버 오류로 처리했습니다.
이 외 에러 코드는 외부 토스 서버에서의 응답과 동일한 httpStatus와 message로 응답하도록 처리했습니다.

summerlunaa replied:
적절히 구분하셨네요. 너무 좋습니다 👍

kjyyjk reviewed (commented):
리뷰 잘부탁드립니다~!

summerlunaa commented on src/main/java/roomescape/payment/PaymentErrorHandler.java:
```diff
@@ -0,0 +1,45 @@
+
+    @Override
+    public void handleError(final ClientHttpResponse response) throws IOException {
+        if (response.getStatusCode().is2xxSuccessful()) {
```
```suggestion
        if (!response.getStatusCode().isError()) {
```
4XX, 5XX만 걸러내려면 이게 더 명확할 수도 있겠네요!

kjyyjk replied:
생각해보니 기존 코드는 3xx번대 응답도 에러로 처리하고 있겠네요..
수정했습니다!

d45ce5951c190314332430ffdeb257c2a9c1560b

summerlunaa commented on src/main/java/roomescape/payment/PaymentErrorHandler.java:
```diff
@@ -0,0 +1,45 @@
+            return;
+        }
+        final String body = new String(getResponseBody(response), StandardCharsets.UTF_8);
+        final PaymentError paymentError = mapper.readValue(body, PaymentError.class);
```
문자열을 객체로 변환할 때 어떤 예외가 발생할 수 있을까요?

kjyyjk replied:
역직렬화하는 과정에서 객체 생성 및 매핑에 문제가 발생하는 경우 Checked Exception인
`JsonProcessingException`(혹은 그 하위 예외인 `JsonMappingException`)이 발생할 수 있습니다.

이 경우에는 서버 개발자만 해당 에러를 확인한 뒤 사용자에게 서버 오류라고 응답해주면 좋을 것 같아요.

현재는 `throws IOException`을 통해 호출부로 예외를 던져주고 있는데, 확인해보니 GlobalExceptionHandler에 IOException, Exception을 처리하는 핸들러가 존재하지 않네요!

따라서 GlobalExceptionHandler에 IOException과 Exception을 처리하는 핸들러를 추가해줬습니다.

b6b21203781ce22ea4972b52cf19b1b86e766ab6

kjyyjk replied:
사실 질문 주신 의도를 잘 파악하지 못한 것 같습니다ㅠ
혹시 잘못 이해하거나 놓친 부분이 있을까요?

summerlunaa replied:
ㅎㅎ 잘 처리해주셨습니다! 다만 `IOException`과 같은 낮은 수준의 예외를 GlobalExceptionHandler로 던지기 보다는, 사용처에서 명확하게 try-catch를 통해 예외를 캐치하여 애플리케이션 수준의 예외로 변환해서 처리하는 것이 더 적절해 보입니다. 이렇게 하면 예외 종류와 메시지를 통해 예외의 의미를 더 명확하게 드러낼 수 있어 개발자가 디버깅하기도 더 쉬워집니다!

summerlunaa commented on src/main/java/roomescape/payment/PaymentErrorHandler.java:
```diff
@@ -0,0 +1,45 @@
+
+        boolean exists = FilteredPaymentErrorCode.exists(paymentError.code());
+        if (exists) {
+            throw new PaymentException(INTERNAL_SERVER_ERROR, "서버 내부 오류입니다.", INTERNAL_SERVER_ERROR.name());
```
서버 내부 오류인 경우에도 개발자는 어떤 오류가 발생했는지 확인할 수 있어야 디버깅이 쉽겠네요 :)

kjyyjk replied:
디버깅할 때마다 항상 GlobalExceptionHandler를 전부 주석처리 했던 기억이 있어, 말씀주신 내용에 매우 동의합니다.
따라서 해결 방안으로 로깅에 대해 학습하고 적용해 볼 생각입니다!

다만 피드백 반영 및 리뷰 재요청이 좀 늦은 것 같아, 이 부분은 다음 리뷰 요청에 포함해도 괜찮을까요?

summerlunaa commented on src/main/java/roomescape/payment/PaymentService.java:
```diff
@@ -0,0 +1,46 @@
+@Service
+public class PaymentService {
+
+    private static final String SECRET_KEY = "test_gsk_docs_OaPz8L5KdmQXkzRz3y47BMw6";
```
secret key, base url(`https://api.tosspayments.com`), timeout 설정들은 yml이나 properties로 분리해볼까요? 분리하면 어떤 이점을 얻을 수 있을까요?

kjyyjk replied:
secret key를 application-auth.yml로, base url과 timeout 설정을 application-pay.yml로 분리했습니다.
(프로필 분리에 따른 추가 설정은 PR 본문에 최신화 해놓겠습니다!)

이렇게 하면 .gitignore에 `application-auth.yml`을 추가해 민감 정보를 git으로 관리하지 않을 수 있다는 이점이 있습니다.
또한 profile들을 조합해 프로덕션 코드 수정 없이 환경마다 다른 설정을 해줄 수 있습니다.

de882c42575df91a1633be36a8cec788df4befd8

summerlunaa replied:
설정값들을 yml로 잘 분리하셨습니다! 그런데 두 yml을 하나로 합쳐주실 수 있을까요? 😅 보안을 위해 auth.yml을 분리한 것은 훌륭하나 현재는 학습용 미션이다보니 제가 환경을 세팅해서 리뷰하기가 번거롭네욥ㅎㅎㅎ 🙇‍♀️

summerlunaa commented on src/main/java/roomescape/payment/PaymentService.java:
```diff
@@ -0,0 +1,46 @@
+import roomescape.reservation.controller.PaymentConfirmResponse;
+
+@Service
+public class PaymentService {
```
서비스가 유지되는 한 결제 기능은 계속 존재하겠지만 연동하는 API는 언제든 변경될 수 있습니다. 토스 페이먼트의 결제 API에서만 한정적으로 사용할 수 있는 설정&클라이언트와 PaymentService를 분리해보면 좋을 것 같아요. 그리고 Toss API에 맞춰진 요청/응답/클라이언트/설정 등에는 다른 API와 구분이 가능하도록 Toss/TossPayment 등을 클래스명에 붙여주는 게 좋겠네요.

kjyyjk replied:
PaymentService가 PaymentClient 인터페이스에 의존하고, TossPaymentClient가 PaymentClient를 구현하게
리팩터링 했습니다.
de882c42575df91a1633be36a8cec788df4befd8

이는 PaymentService가 구현체가 아닌 추상화에 의존하기 때문에, 외부 서버에 대한 변경에 유연하다는 이점을 가집니다.
하지만 객체 지향 설계의 단점과 같이 클래스 수가 늘고 구조가 복잡해진다는 문제가 있습니다.

피드백을 바로 적용한 이유는 방탈출 예약 시스템의 결제사가 변경될 확률과 그때의 변경 비용을 고려해봤을 때,
현재 미리 가져가는 것이 합리적이라고 판단했기 때문입니다!

최근 이러한 추상화 구조(혹은 DIP)를 언제 가져가는 것이 좋을 지에 대한 고민을 자주하는데 참 어려운 것 같아요.ㅠ
듣기로는 개발자마다 기준이 다르다고 하는데, 파랑만의 기준이 있다면 어떤 것인지 궁금합니다!

summerlunaa replied:
잘 변경해주셨습니다 👍  

> 이러한 추상화 구조(혹은 DIP)를 언제 가져가는 것이 좋을 지에 대한 고민

개인적으로는 근시일 내로 변경의 가능성이 명확하거나, 추상화를 통해 당장 테스트 환경을 편하게 구축할 수 있다던가, 추후 추상화된 구조로 변경하는 비용이 크다고 판단되면 초기에 추상화를 하는 편입니다. 
지금 다루고 있는 외부 API 호출은 내부적으로 컨트롤이 어렵기 때문에 비즈니스 로직과의 의존성 분리를 위해 충분히 추상화 할만한 부분이라 생각됩니다. 당장 결제사가 변경되지 않더라도 추상화를 통해 테스트도 편하게 할 수 있고 말씀하신 것처럼 결제사가 변경되는 상황에도 유연하게 대응할 수 있으니까요!

물론 추상화 없이 Service와 Client만 분리하는 것도 현재 요구사항에서는 괜찮은 선택이라고 생각합니다. 사실 당장 다른 결제사로 변경될 예정이 없는 데에다가 현재의 `PaymentConfirmRequest/Response` 구조가 다른 PG사의 client에도 동일하게 적용될 수 있을까 하는 의문이 들기도 해서요. Client와 Service를 분리하는 것만으로도 결제 비즈니스 로직 처리와 외부 client 호출을 분리할 수 있으니 현재로써는 그 정도도 나쁘지 않은 것 같습니다 :)

summerlunaa commented on src/main/java/roomescape/reservation/controller/PaymentConfirmRequest.java:
```diff
@@ -0,0 +1,8 @@
+package roomescape.reservation.controller;
```
패키지의 위치가 dto여야 하겠네요!

kjyyjk replied:
수정했습니다!

summerlunaa commented on src/main/java/roomescape/payment/PaymentService.java:
```diff
@@ -0,0 +1,46 @@
+    private static final String AUTHORIZATION_PREFIX = "Basic ";
+    private static final String AUTHORIZATION_DELIMITER = ":";
+
+    private final RestClient restClient;
```
RestClient를 선택한 이유가 궁금합니다!

kjyyjk replied:
Http Client 중 RestTemplate, RestClient, WebClient 이렇게 세 가지 선택지를 고려해봤습니다.

우선 WebClient는 주로 비동기 방식이 필요할 때 사용한다고 하는데, 현재 요구사항에서 그러한 기술은 필요가 없다고 판단해 선택하지 않았습니다.

따라서 RestTemplate과 RestClient 중 하나를 선택해야했습니다.
정확한 비교를 두 방식으로 간단한 CRUD요청을 구현해보았고, 느낀 점은 다음과 같습니다.
- RestClient는 체이닝 메서드를 지원해 가독성 측면에서 더 좋았습니다.
- RestTemplate, RestClient 모두 ResponseHandler 기능을 제공하나, RestClient는 `onStatus()`를 통해 더 손쉽게 에러를 핸들링할 수 있었습니다.
- RestClient는 `header()` 와 같은 체이닝 메서드를 활용해 손쉽게 헤더를 설정할 수 있었습니다.
반면 RestTemplate은 HttpEntity(혹은 RequestEntity)를 통해 직접 헤더를 설정해주어야 했습니다.

이처럼 사용하기에는 RestClient가 조금 더 편한 느낌이었습니다.

추가로 RestTemplate은 더 이상 신규 기능을 지원하지 않고 유지만 한다는 것을 알 수 있었습니다.
공식 문서에서도 새로 시작한다면 RestClient를 추천하고 있었습니다.

반면 RestClient가 RestTemplate에 비해 자료가 적다는 의견도 있었는데요!
실제로 RestClient를 검색해 자료를 읽다 보면 RestTemplate 관련 자료였던 적도 여럿 있었습니다.

하지만 현재 수준에서는 RestClient의 자료 부족이라는 단점보다 장점이 훨씬 크게 느껴져,
RestClient를 선택했습니다.

summerlunaa commented on src/main/java/roomescape/common/exception/PaymentException.java:
```diff
@@ -0,0 +1,25 @@
+import org.springframework.http.HttpStatusCode;
+
+@Getter
+public class PaymentException extends RuntimeException {
```
적절한 커스텀 예외 사용 좋습니다 👍

summerlunaa commented on src/main/java/roomescape/reservation/controller/ReservationController.java:
```diff
@@ -45,14 +48,19 @@ public List<AvailableReservationTimeWebResponse> getAvailable(
-                createReservationWebRequest,
+                request.toCreateReservationWebRequest(),
                 memberInfo
         );
```
결제 후 예약을 생성하도록 순서를 배치한 이유가 궁금합니다.

kjyyjk replied:
기존에는 결제가 완료된 뒤 예약을 생성하는 순서가 자연스럽다고 생각했었습니다.

하지만 예약 생성 과정에서 예외가 발생하게 된다면 결제 취소를 해야 한다는 점도 고려해야 할 것 같아요.
그런 점에서는 예약을 문제 없이 생성하고 난 뒤에 결제를 하는 것이 합리적이고 효과적이라고 생각됩니다.
(결제 취소보다는 결제 중 에러 발생 시 예약 생성을 롤백하는 것이 더 간단하기도 하니까요.)

순서를 조정하고, 트랜잭션을 적용하기 위해 `ReservationService#createAndPay`를 만들었습니다.

971a7c45c6db65494c5fef0a1856b056091e0c97

summerlunaa reviewed (changes requested):
레오 안녕하세요! 파랑입니다 🐳 만나서 반가워요~~
결제 API 연동이 쉽지 않았을 것 같은데 잘 구현해주셨네요 👍 테스트도 잘 챙겨주셨구요!
몇 가지 질문과 리뷰 남겨두었으니 확인 부탁드립니다.
혹시 어떤 주제든 질문이 있거나 리뷰에서 이해 안 되는 부분 있으면 언제든 편하게 DM 주셔도 괜찮습니다~ 마지막 미션도 파이팅 해보시죠 💪

kjyyjk committed:
refactor: TimeOut 기준 시간 수정

- 내가 사용자라고 상상해봤을 때 ConnectionTimeOut과 ReadTimeOut의 기준 시간 합이
10초를 넘겨서는 안된다고 생각했다. 보통 ConnectionTimeOut이 ReadTimeOut보다
짧으므로 각각 2초, 8초로 설정했다.

kjyyjk committed:
refactor: 결제 에러 처리 로직 리팩터링

kjyyjk committed:
feat: 추가 예외 처리 핸들러 구현

kjyyjk committed:
refactor: 결제 및 예약 로직 리팩터링

- ReservationService#createAndPay로 예약 생성과 결제를 하나의 트랜잭션 내에서 수행한다. 외부 API는 트랜잭션 롤백되지 않으므로 결제는 마지막에 수행한다.

kjyyjk committed:
refactor: 결제 로직 리팩토링

- TossPaymentClient를 PaymentClient로 추상화하고 PaymentService가 PaymentClient를 주입받는
구조로 리팩터링한다. 추후 토스 페이먼츠 API가 아닌 다른 결제 API로 변경되어도 유연하다는 이점이 있다.

- secret key, base url 등 하드 코딩된 설정 정보를 yml로 분리한다.
민감 정보 관리, 환경 별 설정 등에 이점을 가진다.

summerlunaa commented on src/main/java/roomescape/common/config/RestClientConfig.java:
```diff
@@ -1,19 +1,25 @@
 @Configuration
 public class RestClientConfig {
 
+    @Value("${timeout.connection}")
+    private int TIMEOUT_CONNECTION;
+    @Value("${timeout.read}")
+    private int TIMEOUT_READ;
```
모든 client가 동일한 timeout 값을 사용할까요? 고민해보세요 :)

참고) `@ConfigurationProperties`와 `@ConfigurationPropertiesScan`을 사용해서 설정값을 가져올 수도 있습니다.

summerlunaa reviewed (approved):
레오 안녕하세요!
리뷰도 잘 반영해주시고 남겨드린 질문도 열심히 찾아보고 답변 주셨네요 👍 
코멘트 조금 더 남겨드렸지만 고 부분은 다음 단계에서 반영해보시면 될 것 같습니다. 1단계는 빠르게 머지하겠습니다. 고생하셨어요 👏
